testing/bdd/test_ebuttd_style_references.py

```diff
@@ -0,0 +1,67 @@
+from pytest_bdd import scenarios, when, then, given, parsers
+import xml.etree.ElementTree as ET
+
+scenarios('features/styles/ebuttd_style_references.feature')
+
+
+@when(parsers.parse('it contains style "{style_name}"'))
+def when_it_contains_style(test_context, template_dict, style_name):
+    if 'styles' not in template_dict:
+        template_dict['styles'] = list()
+    style = {"id": style_name}    
+    template_dict['styles'].append(style)
+    test_context[style_name] = style
+
+
+@when(parsers.parse('style "{style_name}" has attribute "{attribute}" set to "{value}"'))
+def when_style_has_attribute(test_context, style_name, attribute, value):
+    test_context[style_name][attribute] = value
+
+
+@when(parsers.parse('it contains some text with style "{style_name}"'))
+def when_style_has_attribute(template_dict, style_name):
+    template_dict['text_style'] = style_name
+
+
+@when(parsers.parse('it contains some text with region "{region_name}"'))
+def when_style_has_attribute(template_dict, region_name):
+    template_dict['text_region'] = region_name
+
+
+@when(parsers.parse('it contains region "{region_id}"'))
+def when_it_contains_region(test_context, template_dict, region_id):
+    if 'regions' not in template_dict:
+        template_dict['regions'] = list()
+    region = {"id": region_id}    
+    template_dict['regions'].append(region)
+    test_context[region_id] = region
+
+
+@when(parsers.parse('region "{region_id}" has attribute "{attribute}" set to "{value}"'))
+def when_region_has_attribute(test_context, region_id, attribute, value):
+    test_context[region_id][attribute] = value
+
+
+@then(parsers.parse('the ebu_tt_d document contains style "{style_name}" with attribute "{attribute}" set to "{value}"'))
+def then_converted_document_has_style(test_context, style_name, attribute, value):
+    ebuttd_document = test_context['ebuttd_document']
+    tree = ET.fromstring(ebuttd_document.get_xml())
+    elements = tree.findall('{http://www.w3.org/ns/ttml}head/{http://www.w3.org/ns/ttml}styling/{http://www.w3.org/ns/ttml}style[@{http://www.w3.org/XML/1998/namespace}id="%s"]' %style_name)
+    assert len(elements) == 1
+    assert elements[0].get('{http://www.w3.org/ns/ttml#styling}%s' % attribute) == value
```

Comment:
Strange we can't write this into a single xpath in the call to `findall` - I wonder if it supports that functionality? It should be something very like:

`'{http://www.w3.org/ns/ttml}head/{http://www.w3.org/ns/ttml}styling/{http://www.w3.org/ns/ttml}style[@{http://www.w3.org/XML/1998/namespace}id="%s"]/@{http://www.w3.org/ns/ttml#styling}%s' %style_name %attribute`

Reply:
It can probably be done, but that looks like quite a long and confusing query. By splitting the step up it is easier to read, and we can check that there is exactly one style with that specified ID.
I'm going to try and refactor these lines because they're quite long right now.

Reply:
OK, I actually thought it would be _easier_ to read, but I'm prepared to be told I'm wrong!

Reply:
Might have been easier to read if the namespaces could be declared like `tt:head` instead of including the full namespace ID.

I've updated this file now. What do you think?

Reply:
This looks good. I might still have combined them into a single xpath by preference, but this may be better, practically: the result is that if the element doesn't exist then that will generate a different assertion error from if the attribute doesn't exist, which is useful, potentially.